app/cosi/cosi.go

```diff
@@ -0,0 +1,374 @@
+	}
+	sig, err := sign(file, groupToml)
+	handleErrorAndExit("Couldn't create signature", err)
+	dbg.Lvl3(sig)
+	sigFileName := fileName + ".sig"
+	outFile, err := os.Create(sigFileName)
+	handleErrorAndExit("Couldn't create signature file: ", err)
+	writeSigAsJSON(sig, outFile)
+	dbg.Lvl1("Signature written to: " + sigFileName)
+}
+
+func signString(c *cli.Context) {
+	msg := strings.NewReader(c.Args().First())
+	groupToml := c.GlobalString("servers")
+	sig, err := sign(msg, groupToml)
+	handleErrorAndExit("Couldn't create signature", err)
+	writeSigAsJSON(sig, os.Stdout)
+}
+
+func verifyFile(c *cli.Context) {
+	dbg.SetDebugVisible(c.GlobalInt("debug"))
+	err := verify(c.Args().First(), c.GlobalString("servers"))
+	verifyPrintResult(err)
+}
+
+func verifyString(c *cli.Context) {
+	f, err := ioutil.TempFile("", "cosi")
+	handleErrorAndExit("Couldn't create temp file", err)
+	f.Write([]byte(c.Args().First()))
+	f.Close()
+	sigfile := f.Name() + ".sig"
+	sig, err := ioutil.ReadFile(c.String("signature"))
+	handleErrorAndExit("Couldn't read signature: ", err)
+	err = ioutil.WriteFile(sigfile, sig, 0444)
+	handleErrorAndExit("Couldn't write tmp-signature", err)
+	err = verify(f.Name(), c.GlobalString("servers"))
+	verifyPrintResult(err)
+	os.Remove(f.Name())
+	os.Remove(sigfile)
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+// verifyPrintResult prints out OK or what failed.
+func verifyPrintResult(err error) {
+	if err == nil {
+		dbg.Print("OK: Signature is valid.")
+	} else {
+		dbg.Print("Invalid: Signature verification failed:", err)
+	}
+}
+
+// writeSigAsJSON - writes the JSON out to a file
+func writeSigAsJSON(res *s.ServiceResponse, outW io.Writer) {
+	b, err := json.Marshal(res)
+	if err != nil {
+		handleErrorAndExit("Couldn't encode signature: ", err)
+	}
+	var out bytes.Buffer
+	json.Indent(&out, b, "", "\t")
+	outW.Write([]byte("\n"))
+	if _, err := out.WriteTo(outW); err != nil {
+		handleErrorAndExit("Couldn't write signature", err)
+	}
+	outW.Write([]byte("\n"))
+}
+
+// handleErrorAndExit is a shortcut for all those pesky err-checks
+func handleErrorAndExit(msg string, e error) {
+	if e != nil {
+		dbg.Fatal(os.Stderr, msg+": "+e.Error())
+	}
+}
+
+// sign takes a stream and a toml file defining the servers
+func sign(r io.Reader, tomlFileName string) (*s.ServiceResponse, error) {
+	dbg.Lvl3("Starting signature")
+	f, err := os.Open(tomlFileName)
+	if err != nil {
+		return nil, err
+	}
+	el, err := config.ReadGroupToml(f)
+	if err != nil {
+		return nil, err
+	}
+	if len(el.List) <= 0 {
+		return nil, errors.New("Empty or invalid cosi group file:" +
+			tomlFileName)
+	}
+	dbg.Lvl2("Sending signature to", el)
+	res, err := signStatement(r, el)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// signStatement can be used to sign the contents passed in the io.Reader
+// (pass an io.File or use an strings.NewReader for strings)
+func signStatement(read io.Reader, el *sda.EntityList) (*s.ServiceResponse,
```

Comment:
I propose we have the services defining an API that can be called from any client wanting to use it - like I did in

https://github.com/dedis/cothority/blob/services_skipchain/services/skipchain/client.go
and
https://github.com/dedis/cothority/blob/services_skipchain/services/skipchain/api.go

what do you think?